### PR TITLE
[core] fix RCTHost not retained

### PR DESCRIPTION
**Pull request by Kudo**
# Why

fix blank view on ios bridgeless mode 

# How

unlike RCTBridge which is retained by RCTRootView. the `RCTHost` is retained by `RCTRootViewFactory` and the RCTRootViewFactory is retained by AppDelegate. if we create a local `RCTRootViewFactory.viewWithModuleName:...`, the created RCTHost will be released right away.

this pr tries to retain the RCTHost/RCTBridge by the root view.

# Test Plan

test on https://github.com/brentvatne/blank-bridgeless 

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-modules-core/ios/ReactDelegates/EXReactRootViewFactory.mm
Comment (lukmccall) [inline packages/expo-modules-core/ios/ReactDelegates/EXReactRootViewFactory.mm:64]:
```suggestion
  if (![appDelegate.rootViewFactory isKindOfClass:EXReactRootViewFactory.class]) {
    @throw [NSException exceptionWithName:NSInternalInconsistencyException
                                   reason:@"The appDelegate.rootViewFactory must be an EXReactRootViewFactory instance."
                                 userInfo:nil];
  }
```

**Review by lukmccall — APPROVED** [review]
LGTM, but please wait for a review from @tsapeta

File: packages/expo-modules-core/ios/ReactDelegates/EXReactRootViewFactory.mm
Comment (tsapeta) [inline packages/expo-modules-core/ios/ReactDelegates/EXReactRootViewFactory.mm:85]: 🙉 
I wish we don't have to do this...

**Review by tsapeta — APPROVED** [review]

File: packages/expo-modules-core/ios/ReactDelegates/EXReactRootViewFactory.mm
Comment (Kudo) [reply packages/expo-modules-core/ios/ReactDelegates/EXReactRootViewFactory.mm:85]: let me think more about this. probably we need to change something for upstream.
it works for greenfield apps because typically it's 1-1 mapping for AppDelegate <-> RCTRootViewFactory. however, we have dev-menu and dev-launcher that would have different RCTHost instances other than the one from AppDelegate. maybe we should also have a RCTRootViewFactory inside their host.

Comment (Kudo) [reply packages/expo-modules-core/ios/ReactDelegates/EXReactRootViewFactory.mm]: @tsapeta does that look better to you? https://github.com/expo/expo/pull/27715/commits/63ff5c9f4fecd04a16c07b500c0b8c626a30a037

Comment (Kudo) [reply packages/expo-modules-core/ios/ReactDelegates/EXReactRootViewFactory.mm]: later we could remove the RCTRootViewFactory internal category when https://github.com/facebook/react-native/pull/43528 is available. i'll keep it as-it and remove the internal category later.

**Pull request merged**
